src/executioners/SpectralExecutionerLinearElastic.C

```diff
@@ -0,0 +1,402 @@
+      "to add. Components are XX, YY, ZZ, YZ, XZ, XY");
+  params.addParam<Real>("solver_error", 1.0e-4, "Error for fixed iteration solver");
+
+  return params;
+}
+
+SpectralExecutionerLinearElastic::SpectralExecutionerLinearElastic(
+    const InputParameters & parameters)
+  : SpectralExecutionerBase(parameters),
+    _dt(getParam<Real>("time_step")),
+    _nsteps(getParam<unsigned int>("number_iterations")),
+    _young_modulus(getParam<Real>("young_modulus")),
+    _poisson_ratio(getParam<Real>("poisson_ratio")),
+    _average_factor(getParam<Real>("average_material_factor")),
+    _solver_error(getParam<Real>("solver_error"))
+{
+
+  _initial_strain_tensor.fillFromInputVector(getParam<std::vector<Real>>("global_strain_tensor"));
+
+  _t_current = 0.0;
+}
+
+void
+SpectralExecutionerLinearElastic::fillOutEpsilonBuffer(
+    FFTBufferBase<RankTwoTensor> & epsilon_buffer)
+{
+  const auto & grid = epsilon_buffer.grid();
+  const int ni = grid[0];
+  const int nj = grid[1];
+  const int nk = grid[2];
+
+  std::size_t index = 0;
+
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        epsilon_buffer.realSpace()[index] = _initial_strain_tensor;
+        index++;
+      }
+}
+
+void
+SpectralExecutionerLinearElastic::getGreensFunction(FFTBufferBase<RankFourTensor> & gamma_hat,
+                                                    const RankFourTensor & elasticity_tensor)
+{
+  const auto & grid = gamma_hat.grid();
+  const int ndim = 3;
+
+  const int ni = grid[0];
+  const int nj = grid[1];
+  const int nk = (grid[2] >> 1) + 1;
+
+  std::size_t index = 0;
+
+  const auto & ivec = gamma_hat.kTable(0);
+  const auto & jvec = gamma_hat.kTable(1);
+  const auto & kvec = gamma_hat.kTable(2);
+
+  const std::vector<int> grid_vector = gamma_hat.grid();
+
+  const Complex I(0.0, 1.0);
+
+  // Fill fourth-order Green operator based on homogeneous material properties
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        const std::array<Complex, 3> freq{ivec[freq_x] * I, jvec[freq_y] * I, kvec[freq_z] * I};
+
+        Real lambda0 = _young_modulus * _average_factor * _poisson_ratio /
+                       ((1 + _poisson_ratio) * (1 - 2 * _poisson_ratio));
+        Real nu0 = _young_modulus * _average_factor / (2 * (1 + _poisson_ratio));
+        Real constant = (lambda0 + nu0) / (nu0 * (lambda0 + 2.0 * nu0));
+
+        for (int i = 0; i < ndim; i++)
+          for (int j = 0; j < ndim; j++)
+            for (int k = 0; k < ndim; k++)
+              for (int l = 0; l < ndim; l++)
+              {
+                Complex q_square = freq[0] * freq[0] + freq[1] * freq[1] + freq[2] * freq[2];
+                if (std::abs(q_square) > 1.0e-12)
+                {
+                  gamma_hat.reciprocalSpace()[index](i, j, k, l) =
```

Comment:
it might be best to grab a reference to `gamma_hat.reciprocalSpace()` outside of the loop. 

```
auto & gamma_hat_F = gamma_hat.reciprocalSpace()
```

(or whatever naming convention you want to adopt for recoprocal fields)

Reply:
Would just be saving the function call, right?

Reply:
A function call in a tight inner loop. As that function is implemented in the header file there is a very good chance the compiler will inline it, however we might as well be explicit about it.